fms/models/__init__.py

```diff
@@ -208,6 +206,26 @@ def _is_dp(distributed_strategy):
     return distributed_strategy in {"fsdp", "hsdp", "ddp"}
 
 
+def _load_state_dict_into_model(
+    model, state_dict, device, rank, world_size, checkpoint_format
+):
+    if checkpoint_format == "st":
+        # In this case state_dict contains some extra information:
+        # 1. Each tensor in the state dict is in a meta device to save memory
+        # 2. Each tensor has 3 extra parameters appended to it:
+        #   - st_file: the safetensors checkpoint to load the weights from
+        #   - st_key: the key(s) in the ST checkpoint needed for this weight
+        #   - transform_func: a function that transforms the weights represented
+        #     by st_key in the st_file checkpoint into the FMS weight in the state dict
+        # This is all the information needed to properly load, transform, and
+        # optionally TP partition the weights coming from a ST checkpoint
```

Comment:
could we encapsulate all this in some kind of LazyStateDict class so that it doesn't look any different from a regular state_dict when used? It seems a bit confusing to have a special path and special handling that only applies to safe tensors. I think this makes the control-flow and application (and availability) of sharding strategies less clear.

Reply:
this is a normal state_dict, the only modifications are inside the tensors themselves, as having the weird state_dict broke the adapters